Improve the docs CRD schema version update script

joestringer opened the pull request:
Review commit-by-commit.

Fix a bunch of quality-of-life things in the CRD schema docs generator script
so it runs, correctly, in my env (different git remote) against prerelease
versions like the upcoming v1.10 release, with the latest branch, with a wider
variety of inputs; and allow it to more clearly state what it is doing while it
is generating the CRD schema versions.

Also fix up the script so it will reliably tell you whether you need to bump the
schema version or not, based on the diff in the tree since the most recent
tagged release on the target branch.

Related: #15797

joestringer committed:
docs/crd: Fix git remote detection for CRD schema

Autodetect the git remote for "upstream" using the script from the tree.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
docs/crd: Fix CRD schema compat duplicate version

We now support multiple versions for CRD schemas so this check was
duplicating the CRD versions incorrectly in the table:

    $ git grep -o 'CustomResourceDefinitionSchemaVersion =.*' \
        upstream/v1.10 -- pkg/k8s
    upstream/v1.10:pkg/k8s/apis/cilium.io/v2/register.go:
        CustomResourceDefinitionSchemaVersion = "1.23.1"
    upstream/v1.10:pkg/k8s/apis/cilium.io/v2alpha1/register.go:
        CustomResourceDefinitionSchemaVersion = "1.23.1"

Add a 'uniq' here so we at least don't mess up the table for these
legitimate cases for now. I don't think we have immediate plans to
support different schema versions for the different resource versions.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
docs/crd: Support pre-release versions

When there's not yet any release, the "stable versions" command fails to
find any results in the git tags and due to 'pipefail', the entire
script fails out (silently).

In this case, just echo "" so the script can continue generating the
schema versions.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
docs/crd: Print status to stderr

To give some indication as to progress of the script and whether it
succeeds or not, print some messages to the stderr of the console.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
docs/crd: Support version args in different formats

Use the old "ersion" / "version" trick to make the argument to the
script more flexible so the user doesn't have to remember whether to use
the 'v' or not.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
docs/crd: Only generate 3 CRD schema versions

When generating CRD schema versions for pre-release branches, drop the
oldest release, just like would be present in the final release docs.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
docs/crd: Fix schema version check

The schema version check was previously broken in two ways:
* It only intended to check whether the CRD version was different from
  the most recent release, which would never be true
* Since it checked the output of [[ ! "$( COMPARISON )" ]], it was
  really just checking whether the textual output of the function was
  present rather than using the boolean value of the function.

Fix it by:
1. Checking whether the git diff from the latest tag to the target
   branch has any changes to the CRDs themselves; and
2. Fixing the comparison check to evaluate the boolean function rather
   than the output result of the function.

Signed-off-by: Joe Stringer <joe@cilium.io>

christarazi reviewed (approved):
:rocket:

qmonnet commented on Documentation/check-crd-compat-table.sh:
```diff
@@ -34,7 +34,14 @@ get_stable_branches(){
+   git ls-remote --tags ${remote} v\* \
+       | awk '{ print $2 }' \
+       | grep "${minor_ver}" \
+       | grep -v '\^' \
```
We can get rid of this grep by passing `--refs` to `git ls-remote`.

qmonnet commented on Documentation/check-crd-compat-table.sh:
```diff
@@ -34,7 +34,14 @@ get_stable_branches(){
+       | grep -v '\^' \
+       | grep -v '\-' \
+       | sed 's+refs/tags/++' \
+       | sort -V \
```
We can remove this `sort` by telling `git` to sort for us: `--sort "version:refname"`.

Side note: I personally don't like too much chaining awk/grep/sed when one process can do the job, but `git ls-remote --tags --sort "version:refname" --refs origin | awk '$2 ~ /'${minor_ver}'/ && !($2 ~ /(v|-)/) {sub("refs/tags/","",$2); print $2}'` is maybe not as readable for everyone so :shrug:.

joestringer replied:
I picked up the --sort. Also for `get_rc_tags_for_minor()` we could fold the multiple greps together so I made the change there too.

Personally I find awk unintelligible beyond the basic `{ print $2 }` type stuff, so I'm more inclined to chain single-use binaries together, at least that way I have a chance of debugging or fixing it if it breaks on the input in future. 

Here's the current diff (still going through the feedback):

```
diff --git a/Documentation/check-crd-compat-table.sh b/Documentation/check-crd-compat-table.sh
index 6f24f535937a..081a41818300 100755
--- a/Documentation/check-crd-compat-table.sh
+++ b/Documentation/check-crd-compat-table.sh
@@ -34,19 +34,20 @@ get_stable_branches(){

 get_stable_tags_for_minor(){
    minor_ver="${1}"
-   git ls-remote --tags ${remote} v\* \
+   git ls-remote --refs --tags ${remote} v\* --sort "version:refname" \
        | awk '{ print $2 }' \
        | grep "${minor_ver}" \
-       | grep -v '\^' \
        | grep -v '\-' \
        | sed 's+refs/tags/++' \
-       | sort -V \
    || echo ""
 }

 get_rc_tags_for_minor(){
    minor_ver="${1}"
-   git ls-remote --tags ${remote} v\* | awk '{ print $2 }' | grep "${minor_ver}" | grep -v '\^' | grep '\-' | sed 's+refs/tags/++' | sort -V
+   git ls-remote --refs --tags ${remote} v\* --sort "version:refname" \
+       | awk '{ print $2 }' \
+       | grep "${minor_ver}.*\-" \
+       | sed 's+refs/tags/++'
 }

 select_last_three() {
```

joestringer replied:
Actually, the `--sort` broke the ordering so I had to drop that ~hunk~ one change:

```
--- a/Documentation/concepts/kubernetes/compatibility-table.rst
+++ b/Documentation/concepts/kubernetes/compatibility-table.rst
@@ -14,6 +14,20 @@
 +-----------------+----------------+
 | v1.7.1          | 1.16           |
 +-----------------+----------------+
+| v1.7.10         | 1.18.1         |
++-----------------+----------------+
+| v1.7.11         | 1.18.1         |
++-----------------+----------------+
+| v1.7.12         | 1.18.1         |
++-----------------+----------------+
+| v1.7.13         | 1.18.1         |
++-----------------+----------------+
+| v1.7.14         | 1.18.1         |
++-----------------+----------------+
+| v1.7.15         | 1.18.1         |
++-----------------+----------------+
+| v1.7.16         | 1.18.1         |
++-----------------+----------------+
 | v1.7.2          | 1.17           |
 +-----------------+----------------+
 | v1.7.3          | 1.17           |
@@ -30,18 +44,6 @@
 +-----------------+----------------+
 | v1.7.9          | 1.18.1         |
 +-----------------+----------------+
-| v1.7.10         | 1.18.1         |
...
```

qmonnet replied:
Strange, this also appeared to work on my setup :thinking: but ok

qmonnet commented on Documentation/check-crd-compat-table.sh:
```diff
@@ -49,6 +49,10 @@ get_rc_tags_for_minor(){
 }
 
+select_last_three() {
+    echo "${@}" | cut -d' ' -f2- -
```
This works as long as we are sure to have 4 items in the list. How about:
```suggestion
    echo "${@}" | rev | cut -d' ' -f-3 | rev
```

joestringer replied:
I tried this and it didn't work ;) Instead of reversing all of the characters in the stream, it reversed each individual argument then eliminated the wrong branch. I don't think it's worth getting this perfect.

qmonnet replied:
Agreed that it's not important.

But I'm getting curious now. It works for me, how can it possibly reverse each argument individually on your machine? :thinking: I would understand if the arguments were separated by newlines, but then the `cut` would not work as expected? (rev from util-linux 2.34 / GNU bash, version 5.0.17)

joestringer replied:
Oh, it worked for me on the commandline, but not in the script ;)

I'm sure there's a way to get it to work, I just figured it's not a problem worth solving unless we decide to change the number of releases we are supporting at a given point in time.

joestringer replied:
Actually, if we just don't call this "select_last_three" then it'd be generic to the count and actually do the right thing. I'll rename the function. Maybe "filter_out_oldest" is better.

qmonnet commented on Documentation/check-crd-compat-table.sh:
```diff
@@ -136,7 +136,8 @@ if [[ "$#" -ne 1 ]]; then
 
-release_version="${1}"
+release_ersion="$(echo $1 | sed 's/^v//g')"
+release_version="v$release_ersion"
```
```suggestion
release_version="${1#v}"
```
should do the trick as well? Or am I missing something with this `g` flag at the end, I don't expect the `v` to be present multiple times?

joestringer replied:
Could potentially do the trick. Agree the `g` is not necessary, since the `^v` will only match on v at the start of the stream anyway. I'm not at all familiar with that bashism, and I suspect that the average contributor will have more familiarity with sed rather than this terse approach.

We've used the sed approach multiple times throughout the tree, I'd suggest if we want to switch over to this shell function instead we can change them all at once. I'll skip this here for now.

joestringer replied:
I dropped the 'g' for what it's worth.

qmonnet commented on Documentation/check-crd-compat-table.sh:
```diff
@@ -141,10 +141,18 @@ release_version="v$release_ersion"
 
 create_file ${release_version} "${dst_file}"
 
+last_cilium_release=$(egrep "[ ]${release_version}[ ]" -B2 "${dst_file}" | head -n 1 | awk '{ print $2 }')
```
Could be:
```suggestion
last_cilium_release=$(egrep "[ ]${release_version}[ ]" -B2 "${dst_file}" | awk 'NR == 1 { print $2 }')

qmonnet commented on Documentation/check-crd-compat-table.sh:
```diff
@@ -141,10 +141,18 @@ release_version="v$release_ersion"
+crd_path="examples/crds"
+if [[ -L $crd_path && -d $crd_path ]]; then
+    crd_path="pkg/k8s/apis/cilium.io/client/crds"
+fi
```
Another option:
```suggestion
crd_path=$(realpath --relative-to . "examples/crds")
```

qmonnet reviewed (commented):
Looks good! I have a few suggestions on the bash code, but nothing blocking. Feel free to ignore the suggestions and merge as-is if you prefer.

joestringer committed:
docs/crd: Simplify some operations

Some of the commands involved more commands than necessary, where better
use of some of the existing tools would perform the same job without
needing to run additional processes (which can run dozens of times to
check all of the releases over the past three releases). Simplify some
of these commands.

Suggested-by: Quentin Monnet <quentin@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>
